transport/src/main/java/io/netty/channel/nio/NioIoHandler.java

```diff
@@ -710,7 +687,52 @@ int selectNow() throws IOException {
         }
     }
 
+    // returns true if selectCnt should be reset
+    private boolean selectReturnPrematurely(int selectCnt,
+                                            long currentTimeNanos,
+                                            long startTimeNanos,
+                                            long timeoutMillis) {
+        if (currentTimeNanos - TimeUnit.MILLISECONDS.toNanos(timeoutMillis) >= startTimeNanos) {
+            // The select operation returned after the timeoutMillis elapsed.
+            if (selectCnt > MIN_PREMATURE_SELECTOR_RETURNS && logger.isDebugEnabled()) {
+                logger.debug("Selector.select() returned prematurely {} times in a row for Selector {}.",
+                    selectCnt - 1, selector);
+            }
+            return true;
+        }
+        return false;
+    }
+
+    // returns true if selectCnt should be reset
+    private boolean unexpectedSelectorWakeup(int selectCnt) throws IOException {
+        if (Thread.interrupted()) {
+            // Thread was interrupted so reset selected keys and break so we not run into a busy loop.
+            // As this is most likely a bug in the handler of the user or it's client library we will
+            // also log it.
+            //
+            // See https://github.com/netty/netty/issues/2426
+            if (logger.isDebugEnabled()) {
+                logger.debug("Selector.select() returned prematurely because " +
+                    "Thread.currentThread().interrupt() was called. Use " +
+                    "NioHandler.shutdownGracefully() to shutdown the NioHandler.");
+            }
+            return true;
+        }
+        return false;
+    }
+
+    private boolean selectReturnPrematurelyManyTimes(final int selectCnt) throws IOException {
```

Comment:
This method is not necessarily about premature selector return, I think? We could call it something like `rebuildSelectorIfNecessary` or something like that.